frameworks/keyed/helix/src/demo/state.cljs

```diff
@@ -0,0 +1,67 @@
+(ns demo.state)
+
+(def adjectives ["pretty", "large", "big", "small", "tall", "short", "long", "handsome", "plain", "quaint", "clean", "elegant", "easy", "angry", "crazy", "helpful", "mushy", "odd", "unsightly", "adorable", "important", "inexpensive", "cheap", "expensive", "fancy"])
+(def colours ["red", "yellow", "blue", "green", "pink", "brown", "purple", "brown", "white", "black", "orange"])
+(def nouns ["table", "chair", "house", "bbq", "desk", "car", "pony", "cookie", "sandwich", "burger", "pizza", "mouse", "keyboard"])
+
+(defrecord Data [id label])
+
+(def id 0)
+
+(defn build-data [count]
+  ;; eagerly build data
+  (loop [i 0
+         data #js []]
+    (if (< i count)
+      (recur
+       (inc i)
+       (doto data
+         (.push (->Data
+                 (set! id (inc id))
+                 (str (rand-nth adjectives)
+                      " "
+                      (rand-nth colours)
+                      " "
+                      (rand-nth nouns))))))
+      ;; done
+      (vec data))))
+
+(defn add [data]
+  (let [last-id (:id (last data) 0)]
+    (into data (build-data 1000))))
+
+(defn update-some [data]
+  (reduce (fn [data index]
+            (let [row (get data index)]
+              (assoc data index (assoc row :label (str (:label row) " !!!")))))
+          data
+          (range 0 (count data) 10)))
+
+(defn swap-rows [data]
+  (if (> (count data) 998)
+    (-> data
+        (assoc 1 (get data 998))
+        (assoc 998 (get data 1)))
```

Comment:
`data[998]` would not change after this, right?

Reply:
I haven't changed this from the previous implementation. `data` is still an immutable vector, so `(get data 1)` will still be the old value.

Reply:
Oh right, I'm obvously affected by the imperative style thinkging here :P